---
title: "recountmethylation Users Guide"
author:
- Sean K. Maden
- Reid F. Thompson
- Kasper D. Hansen
- Abhi Nellore
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document: 
    toc: yes
    toc_depth: 2
  BiocStyle::html_document:
    code_folding: show
    toc: yes
    tocfloat: yes
  html_document:
    df_print: paged
    toc: yes
bibliography: bibliography.bib
package: recountmethylation
abstract: |
  This is the main vignette for the `recountmethylation` package.
vignette: "%\\VignetteDepends{RCurl} %\\VignetteIndexEntry{bioc_vignette}  %\\usepackage[UTF-8]{inputenc}
  %\\VignetteEngine{knitr::knitr} \n"
---

```{r setup, echo = FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE, warning = FALSE, message = FALSE)
```

# Introduction

The `recountmethylation` package enables convenient access to compilations of DNA methylation (DNAm) assay data and metadata available on the data server (https://recount.bio/data/). Compilations contain data from over 35,000 samples obtained from the Gene Expression Omnibus (GEO). The 4 main data files may be downloaded using your browser or by specifying `which.dn` in the `get_rmdl` function. 

This vignette shows how to use `recountmethylation` with 2 small example files. Some background about DNAm array measures and data management with `SummarizedExperiment` objects is assumed.

## Disclaimer

```{r disclaimer, message = TRUE, eval = TRUE}
library(recountmethylation)
```

## Data sources and compilations

Data is from bulk human tissues and cell lines available in the Gene Expression Omnibus as of March 2019. Samples were run on Infinium HM450K arrays, a popular 2-channel array platform that uses BeadArray technology to probe over 480,000 CpG loci genome-wide. These probes have enriched coverage at CG islands, genes, and enhancers (@sandoval_validation_2011).

For a given sample, the array reader generates a pair of intensity files (IDATs), one each for the red and green color channels. In addition to BeadArray probe signals, IDATs contain control probe signals for quality evaluations @noauthor_illumina_2010. These paired channel tables are typically read into `SummarizedExperiment` objects and processed with packages like `minfi` and `ChAMP`.

DNAm channel intensities, methylated and unmethylated signals, and DNAm levels or percentages were stored in an HDF5 database, then data subsets were piped into 3 `HDF5-SummarizedExperiment` objects. All 4 files are accessible from the server (recount.bio/data). While most users will find the `HDF5-SummarizedExperiment` files the most convenient to work with from an R session, `recountmethylation` also includes accessors for the HDF5 database.

## Sample and study metadata

Sample characteristics are included in provided metadata. To retrieve sample metadata, use `get_mdpost(path)` for the HDF5 files. For `HDF5-SummarizedExperiment` files, use `minfi::pData(object)` where `object` points to the loaded file path. Metadata includes the GEO record identifiers for samples (GSM) and studies (GSE), sample record titles, learned labels for tissue and disease, sample type predictions, and model-based predictions. For details, refer to the cited manuscript (@maden_human_2020).

## HDF5 database

HDF5, or hierarchical data format 5, is convenient for storage and access of matrices with tall-and-wide dimensions and non-sparse characteristics, including tables of array-wide DNAm signals for tens of thousands of samples. HDF5 datasets contain homogeneous data (e.g. all variables share the same format), and they use chunking and compression to expedite queries and reduce disk occupancy. An HDF5 database containing DNAm data and metadata is provided in a single 120Gb `.h5` file. Queries, summaries, and handling are supported by the `data_mdpost` and `getrg` functions. The `rhdf5` and `HDF5` packages provide additional support for handling HDF5 datasets.

HDF5 `.h5` file details:

* `remethdb2.h5` (class: `HDF5`, size: 120 Gb)

## HDF5-SummarizedExperiment files

The 3 provided `HDF5-SummarizedExperiment` files represents distinct types of large `SummarizedExperiment` objects. The `RGChannelSet` is the most comprehensive and includes both control and genome-mapping red and green channel intensities. The `GenomicRanges` object contains methylated and unmethylated signals, and the `GenomicMethylSet` set includes DNAm levels or percentages. 

Most users will prefer these files to the HDF5 database for use in an R environment. Each uses a DelayedArray backend that allows rapid data summaries and subsetting and more natural object-oriented scripting without the need to manage database connectors. The file structure is a directory containing a large `assays` file and a small `se.rds` file.

`HDF5-SummarizedExperiment` file details:

* `remethdb_h5se_gm_00-00-01_1583780004` (class: `GenomicMethylSet`, size: 86 Gb)
* `remethdb_h5se_gr_00-00-01_1583780004` (class: `GenomicRanges`, size: 132 Gb)
* `remethdb_h5se_rg_00-00-01_1583780004` (class: `RGChannelSet`, size: 119 Gb)

Additional guidance for analyses can be found in the `data_analysis` vignette.

The examples below show how to retrieve the file path, perform data summaries, query samples and subset the data, and perform a brief analysis for 2 small example files, an HDF5 `.h5` file and a `HDF5-SummarizedExperiment` file.

# HDF5-SummarizedExperiment example

## Getting the file path 

Obtain the file path and instantiate a pointer object using `loadHDF5SummarizedExperiment`.

```{r fpath_h5se, eval = TRUE}
library(HDF5Array)
fn <- "remethdb_h5se-test_gr_00-00-01_1583780004"
h5se.path <- system.file("extdata", fn, package = "recountmethylation")
h5setest <- loadHDF5SummarizedExperiment(h5se.path)
```

The object `h5setest` points to the on-disk dataset, and can now be treated as if it were in active memory. Note the class registers simply as `r class(h5setest)[1]`.

Note the contents of the data directory for this file. 

```{r, eval = TRUE}
list.files(h5se.path)
```

Directories lacking full versions of these files won't be properly recognized by `loadHDF5SummarizedExperiment()`. If this results from calling `get_rmdl`, this may indicate the download failed.

## Data summaries

Inspect part of the dataset using standard functions like `class`, `dim`, and `summarize`.

```{r}
dinfo <- function(ds, name){
  lr <- list("class" = class(ds), "dim" = dim(ds),
             "colnames.head" = head(colnames(ds)), 
             "rownames.head" = head(rownames(ds)),
             "summary" = summary(ds))
  return(lr)
}
dinfo(h5setest)
```

Using functions from `minfi`, obtain metadata with `pData`, DNAm fraction matrices with `getBeta`, and annotation with `getAnnotation`, then inspect them.

```{r}
library(minfi)
ld <- list("metadata" = pData(h5setest),
           "dnam" = getBeta(h5setest),
           "anno" = getAnnotation(h5setest))
for(d in 1:length(ld)){
  print(paste0("FILE: ", names(ld)[d]))
  print(dinfo(ld[[d]]))
}
```

## Sample queries and subsets

## Analysis

# HDF5 example

## Getting the file path

Obtain the path to the test `.h5` file.

```{r fpath_h5}
path <- system.file("extdata", "testh5", package = "recountmethylation")
h5fn <- list.files(path)
h5.path <- paste0(path, "/", h5fn)
```

## Data summaries

The HDF5 file can be summarized with the `h5ls` function from the `rhdf5` package.

```{r}
library(rhdf5)
h5ls(h5.path)
```

Pass the file path to `data_mdpost` and `getrg` to obtain the metadata and the signal data as an RGChannelSet, respectively.

```{r}
pd1 <- data_mdpost(h5.path)
rg <- getrg(dbn = h5.path, all.gsm = TRUE, verbose = TRUE)
pd2 <- pData(rg)
```

Finally, inspect these objects as in the prior example.

```{r}
ld <- list("metadata.pd1" = pd1, "metadata.pd2" = pd2,
           "dnam" = getBeta(rg), "anno" = getAnnotation(rg))
for(d in 1:length(ld)){
  print(paste0("FILE: ", names(ld)[d]))
  print(dinfo(ld[[d]]))
}
```

Note the metadata objects pd1 and pd2 should be identical.

```{r}
identical(pd1, pd2)
```

# Validation

As noted in the disclaimer, it is useful to check mined data with the latest published version, where possible. This package provides the `gds_idat2rg` function, which takes a GSM ID vector, downloads IDATs from GEO, and returns an RGChannelSet.

## New IDAT downloads

For the samples "" and "" in the test dataset above, obtain the data as follows.

```{r}
gsmv <- c()
rgpath <- gds_idat2rg(gsmv) # retrieve idats as RGChannelSet
```

## Data comparisons

Then validate the test data by comparing.

```{r}
```

For additional resources to validate data from GEO, please see the `GEOmetadb` and `GEOquery` packages. 

# Session info

```{r get_sessioninfo}
sessionInfo()
```
