---
title: "recountmethylation Users Guide"
author:
- Sean K. Maden
- Reid F. Thompson
- Kasper D. Hansen
- Abhinav Nellore
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document: 
    toc: yes
    toc_depth: 2
  BiocStyle::html_document:
    code_folding: show
    toc: yes
    tocfloat: yes
  html_document:
    df_print: paged
    toc: yes
bibliography: bibliography.bib
package: recountmethylation
vignette: "%\\VignetteDepends{RCurl} %\\VignetteIndexEntry{bioc_vignette}  %\\usepackage[UTF-8]{inputenc} %\\VignetteEngine{knitr::knitr}\n"
---

```{r setup, echo = FALSE}
suppressMessages(library(knitr))
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE)
```

# Introduction

The `recountmethylation` package enables convenient access to databases of sample assays and metadata from over 35,000 GSM records in the Gene Expression Omnibus (GEO). Database files are offered in 2 principal types, either `HDF5-SummarizedExperiment` or `HDF5`. This User's Guide describes the database files, formats, and classes, and demonstrates basic data acquisition and handling with 2 test files. Some background about DNAm arrays, DNAm measurement, and `SummarizedExperiment` objects is assumed. Consult the Data Analyses vignette for more analysis examples.

## Disclaimer

```{r disclaimer, echo = FALSE, message = TRUE}
library(recountmethylation)
```

## DNAm arrays and workflows background

Databases include human samples run on the Illumina Infinium HM450K BeadArray platform. HM450K is a popular 2-channel platform that probes over 480,000 CpG loci genome-wide, with enriched coverage at CG islands, genes, and enhancers (@sandoval_validation_2011). 

Array processing generates 2 intensity files (IDATs) per sample, one each for the red and green color channels. These raw files also contain control signals useful for quality evaluations @noauthor_illumina_2010. 

## Database files

Database files include 1 `h5` file and 3 `HDF5-SummarizedExperiment` directories. The `HDF5-SummarizedExperiment` files combine the benefits of `HDF5` and `SummarizedExperiment` for handling large datasets, and they use DelayedArray backend technology to allow rapid data summaries. These will be the most convenient data type to work with for R users. The HDF5 database is also provided, and access is supported in `recountmethylation`. HDF5, short for hierarchical data format 5, combines compression and chunking to handle large datasets. The `h5` database file contains red and green signal information and sample metadata.

## SummarizedExperiment classes

Sample IDATs can be read into an R session as an object of class `RGChannelSet`, a type of `SummarizedExperiment`. This object class contains slots for assay matrices, sample metadata, and experiment metadata for DNAm array platforms. Typical workflows convert `RGChannelSet` data into new objects classes like `MethylSet` or `RatioSet` as a result of preprocessing, normalization, etc. While, not all IDAT information is accessible from every object class, the tradeoff is that more processed objects may be smaller and/or faster to access.

Access has been provided to 3 classes of SummarizedExperiment. First, an unnormalized `RGChannelSet` including red and green signals is provided. Second, an unnormalized `MethylSet`, which includes methylated and unmethylated signals. Finally, a normalized `GenomicRatioSet` that contains DNAm fractions. Normalization used the out-of-band signal, or "noob", method, a type of within-sample normalization effective for removing signal technical artifacts (triche_low-level_2013).

## Database files and downloads

The latest available database files are as follows.

```{r, echo = FALSE}
url = "https://recount.bio/data/"
sslver <- FALSE
ftpuseopt <- FALSE
dirlistopt <- FALSE
dn <- RCurl::getURL(url, ftp.use.epsv = ftpuseopt, dirlistonly = dirlistopt,
                    .opts = list(ssl.verifypeer = sslver))
sm <- as.data.frame(servermatrix(dn, sslver), stringsAsFactors = FALSE)
sdf <- as.data.frame(sm)
tsv <- as.numeric(gsub(".*_", "", sdf[,1]))
sdff <- sdf[which(tsv == max(tsv, na.rm = TRUE)),]
kable(sdff)
```

Note you will need between 80-120Gb of free space to store a single compilation file.

Database downloads are managed entirely from `recountmethylation` with the `get_db` set of functions. These manage downloads and access for the latest database files.  See `?get_db` for more info.

## Sample and study metadata

Sample and study metadata tables are included with DNAm data. These contain GEO record IDs for samples (GSM) and studies (GSE), record titles, learned labels for tissue and disease, sample type predictions, model-based predictions, etc (see @maden_human_2020 for details). To access sample metadata from a database file, see the examples below.

# HDF5-SummarizedExperiment files and example

## Obtain the test database

```{r}
h5se.test <- getdb_h5se_test()
```

## Inspect and summarize the database

Inspect the dataset using standard functions like `class`, `dim`, and `summary`.

```{r}
class(h5se.test)
```

```{r}
dim(h5se.test)
```

```{r}
summary(h5se.test)
```

Functions from the minfi package can help to access various types of information. Access the sample metadata using `pData`. 

```{r}
require(minfi)
h5se.md <- pData(h5se.test)
dim(h5se.md)
```
```{r}
colnames(h5se.md)
```

Access CpG-specific DNAm fractions, or "Beta-values", with `getBeta`.

```{r}
h5se.bm <- getBeta(h5se.test)
dim(h5se.bm)
```
```{r}
colnames(h5se.bm) <- h5se.test$gsm
kable(head(h5se.bm))
```

Finally, access the manifest information, including genome coordinates and region info, using `getAnnotation`.

```{r}
an <- getAnnotation(h5se.test)
dim(an)
```

```{r}
colnames(an)
```

```{r}
kable(head(an))
```

# HDF5 database and example

## Obtain the test database

Download the test `h5` file and get its path.

```{r, eval = FALSE}
h5.test <- getdb_h5_test()
```

## Inspect and summarize the database

Read the datasets into an `RGChannelSet` using `getrg`.

```{r, eval = FALSE}
h5.rg <- getrg(dbn = h5.test, all.gsm = TRUE)
```

As above, use `pData` and `getAnnotation` to get sample metadata and array manifest information. Since this is an `RGChannelSet` object, green and red signals are accessible using `getRed` and `getGreen`.

```{r, eval = FALSE}
h5.red <- getRed(rg.h5)
h5.green <- getGreen(rg.h5)
dim(h5.red)
```
```{r, eval = FALSE}
kable(head(reds))
```
```{r, eval = FALSE}
kable(head(greens))
```
```{r, eval = FALSE}
identical(rownames(reds), rownames(greens))
```

Note that signal matrices have bead address identifiers instead of the probe cg ids seen above. 

```{r, eval = FALSE}
head(rownames(reds))
```

# Validate DNAm datasets

As the disclaimer notes, it is good practice to validate the compiled data against the latest available files from GEO. This section shows how to use `recountmethylation` to obtain DNAm array data from GEO and compare with the example data.

## Download and read IDATs from GEO

Obtain a new `RGChannelSet` from downloaded GEO IDATs with `gds_idat2rg`, specifying the test dataset GSM IDs in the `gsmv` argument.

```{r}
# gsmv <- c(colnames(h5.test), colnames(h5se.test))
gsmv <- gsub("\\..*", "", colnames(h5se.test))
geo.rg <- gds_idat2rg(gsmv)
```

## Compare DNAm signals

Compare unnormalized the red and green signal.

```{r, eval = FALSE}
geo.red <- getRed(geo.rg)
geo.green <- getGreen(geo.rg)
```
```{r, eval = FALSE}
identical(geo.red, h5.red)
```
```{r, eval = FALSE}
identical(geo.green, h5.green)
```

## Compare DNAm Beta-values

Next, compare GEO and test data noob-normalized DNAm Beta-values. Note that noob is a within-sample normalization, and so we can subset to the samples of interest after `preprocessNoob`.

```{r}
geo.gr <- preprocessNoob(geo.rg)
colnames(geo.gr) <- gsub("_.*", "", colnames(geo.gr))
colnames(h5se.test) <- gsub("\\..*", "", colnames(h5se.test))
```
```{r}
geo.bm <- as.matrix(getBeta(geo.gr))
h5se.bm <- as.matrix(h5se.bm)
geo.bm <- geo.bm[rownames(geo.bm) %in% rownames(h5se.bm),]
geo.bm <- geo.bm[order(match(rownames(geo.bm), rownames(h5se.bm))),]
```
```{r}
identical(rownames(geo.bm), rownames(h5se.bm))
identical(colnames(geo.bm), colnames(h5se.bm))
```
```{r}
identical(geo.bm, h5se.bm)
```

# Conclusions

This User's Guide provided background about the database files accessed with `recountmethylation`, and provided basic handling, summarization, and validation examples for each of the principal database file types. Consult the Data Analyses vignette and the manuscript @maden_human_2020 for analysis examples and details about data compilations.

# Session info

```{r get_sessioninfo}
sessionInfo()
```

# Works Cited
